// dal/index.js
"use strict";

var MongoClient = require('mongodb').MongoClient;
var ObjectID = require('mongodb').ObjectID;

var connectionString = "";
var database = "";

module.exports.startMongo = function (dbConfig, callback) {
  connectionString = "mongodb://" + dbConfig.host + ":" + dbConfig.port + "/" + dbConfig.name;
  MongoClient.connect(connectionString, function (err, db) {
    database = db;
    callback(err, database);
  });
};

module.exports.insertDocument = function (collectionName, document, callback) {
  console.dir("Inserting into Collection:" + collectionName);

  database.collection(collectionName).insert(document, function (err, docs) {
    callback(err, docs);
  });
};

module.exports.updateDocument = function (collectionName, document, callback) {
  console.dir("Collection Name:" + collectionName);
  document._id = new ObjectID(document._id.toString());

  database.collection(collectionName).update({"_id": document._id}, document, {upsert: true}, function (err, docs) {
    callback(err, docs);
  });
};

module.exports.deleteDocument = function (collectionName, document, callback) {
  console.dir("Collection Name:" + collectionName);

  document._id = new ObjectID(document._id);
  database.collection(collectionName).remove({"_id": document._id}, document, {upsert: false}, function (err, docs) {
    callback(err, docs);
  });
};

module.exports.getAllFromCollection = function (collectionName, callback) {
  console.dir("Collection Name:" + collectionName);

  database.collection(collectionName).find().toArray(function (err, docs) {
    console.dir("docs fetched");
    //console.dir(docs);
    callback(err, docs);
  });
};

module.exports.getTopXFromCollection = function (collectionName, limit, callback) {
  console.dir("Collection Name:" + collectionName);
  
  database.collection(collectionName).find().limit(limit).toArray(function (err, docs) {
    console.dir("docs fetched");
    //console.dir(docs);
    callback(err, docs);
  });
};

module.exports.loadSelectedFields = function (collectionName, query, option, callback) {
  console.dir("Collection Name:" + collectionName);

  database.collection(collectionName).find({}, option).toArray(function (err, docs) {
    console.dir("docs fetched");
    //console.dir(docs);
    callback(err, docs);
  });
};

module.exports.findOne = function (collectionName, query, callback) {
  console.dir("Collection Name:" + collectionName);

  database.collection(collectionName).findOne(query, function (err, docs) {
    console.dir("docs fetched");
    //console.dir(docs);
    callback(err, docs);
  });
};

module.exports.findAll = function (collectionName, query, callback) {
  console.dir("Collection Name:" + collectionName);

  database.collection(collectionName).find(query).toArray(function (err, docs) {
    console.dir("docs fetched");
    //console.dir(docs);
    callback(err, docs);
  });
};
